packages/lib-classifier/src/store/AnnotationsStore.js

```diff
@@ -17,6 +17,20 @@ const AnnotationsStore = types
     }
   }))
   .actions(self => {
+    function afterAttach() {
+      createClassificationObserver()
+    }
+
+    function createClassificationObserver() {
+      const classificationsDisposer = autorun(() => {
+        const classifications = getParent(self)
+        onAction(classifications, (call) => {
+          if (call.name === 'reset') self.reset()
```

Comment:
I think `onAction` is expected to call either `next()` or `abort()` after interrupting an action.

Reply:
It's not interrupting an action. `onAction` just provides a means for adding a callback for actions: https://mobx-state-tree.js.org/API/#onaction

I think you're thinking of the handler for custom middleware: https://mobx-state-tree.js.org/concepts/middleware

Reply:
I am! My bad.

So this calls `classifications.active.annotations.reset()` before `classifications.active` is cleared. Why doesn't clearing `classifications.active` clear out the annotations. It would be good to write tests for that before adding any new code.

Reply:
`classification.active` is an unique reference to a resource stored in the the map of classification resources. When the reference is invalidated by setting it to undefined, then what MST does is remove the resource it references from the map. I don't believe it'll taken any action related to any other node in the tree, even a child node, and I think that's for safety purposes. You don't want cascading side effects that you're unaware of I think. So what we have happening now is we have an orphaned Annotations node and thus the errors.

Reply:
Ah right, so when we invalidate a node, we should invalidate its children first? This is definitely worth having a test for.

By the way, [3.14.1](https://github.com/mobxjs/mobx-state-tree/blob/master/changelog.md#3141) added `acceptsUndefined` to maps. I've been wondering if that's why we saw behaviour change in 3.15.

Reply:
Good to know about the change. Definitely think that's plausible when we attempted the update.